### PR TITLE
Fetch playbook

### PR DESCRIPTION
**Pull request by daryakoval**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
demisto/etc#34585

## Description
Added playbook and script.

## Minimum version of Cortex XSOAR
- [x] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    """
+    Args:
+        args(dict): args from demisto
+
+    Returns: Command Results with context and human readable output
+    """
+    fields_to_search = args.get('fields_to_search', '')
+    context = args.get('context', '{}')
+
+    fields_to_search_array = [field.strip() for field in fields_to_search.split(',')]
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.py:54]:
```suggestion
    fields_to_search = args.get(argToList('fields_to_search', ''))
    context = args.get('context', '{}')
```

File: Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+    if not fields_to_search:
+        raise ValueError('fields_to_search not specified')
+    if not context:
+        raise ValueError('context not specified')
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.py:59]: This is redundant as you set the input as required

File: Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+- name: context
+  required: true
+  description: Map to search fields in
+  isArray: true
+comment: Receive fields and context. Verifies that these field exists in context.
+commonfields:
+  id: VerifyContextFieldsList
+  version: -1
+enabled: false
+name: VerifyContextFieldsList
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.yml:15]: I think the name should be VerifyObjectFieldsList as the script alone has no relation to context.

File: Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+args:
+- name: fields_to_search
+  required: true
+  description: Fields to search if there are exsist in map
+  isArray: true
+- name: context
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyContextFieldsList/VerifyContextFieldsList.yml:6]:
```suggestion
- name: object
```

File: Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
+
+
+def main():
+    args = demisto.args()
+    demisto.info('hi1')
+    query = args.get('query')
+    size = int(args.get('size'))
+    demisto.info('hi')
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py:10]:
```suggestion
    args = demisto.args()
    query = args.get('query')
    size = int(args.get('size'))
```

File: Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+    raw_result = demisto.executeCommand("SearchIncidentsV2", {"query": query,
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py:12]: Need to check for failure here

File: Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+                                                              "size": size})
+    incidents_len = len(raw_result[0].get("Contents", [""])[0].get("Contents", {}).get("data"))
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.py:14]:
```suggestion
    incidents_len = len(raw_result[0].get("Contents", [{}])[0].get("Contents", {}).get("data"))
```

File: Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+commonfields:
+  id: VerifyEnoughIncidents
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyEnoughIncidents/VerifyEnoughIncidents.yml:2]: Don't see this being used anywhere what do I miss?

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+def main():
+    try:
+        return_results(health_check_command(demisto.args()))
+    except Exception as ex:
+        demisto.error(traceback.format_exc())  # print the traceback
+        return_error(f'Failed to execute Script. Error: {str(ex)}')
+
+
+''' ENTRY POINT '''
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py:53]:
```suggestion
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            return False if integration.get('lastError') else True
+    return False
+
+
+def health_check_command(args: Dict[str, Any]) -> CommandResults:
+
+    integration_name = args.get('integration_name', '')
+    if not integration_name:
+        raise ValueError('integration_name not specified')
+
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py:21]: This is required

```suggestion
    integration_name = args.get('integration_name')
```

**Comment by daryakoval** [comment]
Please see the following explanation [here](https://github.com/demisto/etc/issues/34585#issuecomment-878271297).

image
 
First lets delete context and `get indicator field to search` using demisto rest api, in order to get searchfield, sourcebrand fields.

Then `Set indicators search query` just set context key: searchQuery, value : sourceBrands:"${serverConfig.response.sysConf.sourcebrand}" (it is the key from server_keys). 

Now we can start checking if feed already fetched indicators. `VerifyEnoughIndicators` and `Check If Indicators exsist` takes care of it.  If fetch didn't happened yet, we will use GenericPolling: 
image


image

Now we use `findIndicators` to get all indicators. 
Then I wrote script in order to check if all fields that we asked are mapped in all Indicators that were fetched. It happends using script `VerifyContextFieldsList` at step named: `Check If Fields Exists`.

After all this, we can check Integration Health using script `VerifyIntegrationHealth` that I wrote too, it checks if there were some errors during the fetch.

File: Packs/DeveloperTools/ReleaseNotes/1_1_8.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+
+#### Scripts
+##### New: VerifyIntegrationHealth
+- Heath check for integration (Available from Cortex XSOAR 6.0.0).
+##### New: VerifyEnoughIncidents
+- Returns, whether there're, are enough incidents in the system. (Available from Cortex XSOAR 6.0.0).
+##### New: VerifyContextFieldsList
+- Receive fields and context. Verifies that these field exists in context. (Available from Cortex XSOAR 6.0.0).
```
Comment (ShahafBenYakir) [inline Packs/DeveloperTools/ReleaseNotes/1_1_8.md:8]:
```suggestion
- Check integration for existing errors (Available from Cortex XSOAR 6.0.0).
##### New: VerifyEnoughIncidents
- Check whether a given query returns enough incidents. (Available from Cortex XSOAR 6.0.0).
##### New: VerifyContextFieldsList
- Verify that a given object includes all the given fields. (Available from Cortex XSOAR 6.0.0).
```

**Review by ShahafBenYakir — COMMENTED** [review]
Initial brief review

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  required: true
+  secret: false
+comment: 'Check integration for existing errors.'
+commonfields:
+  id: VerifyIntegrationHealth
+  version: -1
+enabled: false
+name: VerifyIntegrationHealth
+outputs:
+- contextPath: IntegrationHealth.isHealth
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml:15]:
```suggestion
- contextPath: IntegrationHealth.isHealthy
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml [hunk 13]
```diff
@@ -0,0 +1,34 @@
+args:
+- default: false
+  description: 'Integration name to check is health'
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml:3]:
```suggestion
  description: 'Integration name to check its health status.'
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  isArray: false
+  name: integration_name
+  required: true
+  secret: false
+comment: 'Check integration for existing errors.'
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml:8]:
```suggestion
comment: 'Checks for existing errors in a given integration.'
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+commonfields:
+  id: VerifyIntegrationHealth
+  version: -1
+enabled: false
+name: VerifyIntegrationHealth
+outputs:
+- contextPath: IntegrationHealth.isHealth
+  description: 'Checks if the given integration got errors.'
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml:16]:
```suggestion
  description: 'Determines the health status of the integration.'
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+  type: Boolean
+- contextPath: IntegrationHealth.fetchDone
+  description: 'Checks if the given integration finished it fetch.'
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.yml:19]:
```suggestion
  description: 'Determines whether the fetch-indicators command was ended.'
```

File: Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    raw_result = demisto.executeCommand(
+        "demisto-api-post",
+        {
+            "uri": "/settings/integration/search",
+            "body": {
+                "size": 10,
+                "query": "name:" + integration_name
+            },
+        })
+    health_dict = raw_result[0]["Contents"]["response"]["health"]
```
Comment (dantavori) [inline Packs/DeveloperTools/Scripts/VerifyIntegrationHealth/VerifyIntegrationHealth.py:29]: please wrap this with error handling: use `is_error()` and `get_error()` CSP functions (also in other uses of executeCommand in the PR).

**Review by dantavori — CHANGES_REQUESTED** [review]
very nice 👍🏽 see my comments.

File: Packs/DeveloperTools/Scripts/VerifyObjectFieldsList/VerifyObjectFieldsList.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+        fields_to_search_array(list): list of fields to search
+        context_json: context to check the fields in
+
+    Returns: True if all fields are in context_json, else false.
+    """
+    try:
+        for fields in fields_to_search_array:
+            components = fields.split('.')
+            new_context = context_json
+            check_components(components, new_context)
```
[the remark this answers is not in the document]
Comment (daryakoval) [reply Packs/DeveloperTools/Scripts/VerifyObjectFieldsList/VerifyObjectFieldsList.py:37]: dt will check if we have at least one indicator that has such field. My script will check that all indicators has such field.

**Comment by daryakoval** [comment]
@dantavori @ShahafBenYakir  
Run Fetch Indicators Test Playbook for 3 different feeds, how can I add integration names to each test? It will be easier to distinguish different feeds.
image

**Review by dantavori — COMMENTED** [review]
presenting the integrations names in the build is possible but it will require some infra changes, not required for now but we can talk about it.
we can make a short call to go over the leftovers, and let's also schedule a demo for tomorrow.

**Review by dantavori — APPROVED** [review]
awesome

**Comment by ShirleyDenkberg** [comment]
@dantavori @ShahafBenYakir Doc review completed.  
@daryakoval There is no README for the VerifyObjectFieldsList script.  Was that an oversight?

**Pull request merged**